domain/credential/service/service.go

```diff
@@ -124,3 +191,183 @@ func (s *Service) WatchCredential(ctx context.Context, tag names.CloudCredential
+// but before validation can complete.
+func (s *Service) CheckAndUpdateCredential(ctx context.Context, id credential.ID, cred cloud.Credential, force bool) ([]UpdateCredentialModelResult, error) {
+	if s.validationContextGetter == nil {
+		return nil, errors.New("cannot validate credential with nil context getter")
+	}
+
+	models, err := s.modelsUsingCredential(ctx, id)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	modelCloud, err := s.st.GetCloud(ctx, id.Cloud)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		modelsErred  bool
+		modelsResult []UpdateCredentialModelResult
+	)
+	for uuid, name := range models {
+		result := UpdateCredentialModelResult{
+			ModelUUID: uuid,
+			ModelName: name,
+		}
+		result.Errors = s.validateCredentialForModel(uuid, modelCloud, id, &cred)
+		modelsResult = append(modelsResult, result)
+		if len(result.Errors) > 0 {
+			modelsErred = true
+		}
+	}
+	// Since we get a map above, for consistency ensure that models are added
+	// sorted by model uuid.
+	sort.Slice(modelsResult, func(i, j int) bool {
+		return modelsResult[i].ModelUUID < modelsResult[j].ModelUUID
+	})
+
+	if modelsErred && !force {
+		return modelsResult, credentialerrors.CredentialValidation
+	}
+
+	existingInvalid, err := s.st.UpsertCloudCredential(ctx, id.Name, id.Cloud, id.Owner, cred)
+	if err != nil {
+		if errors.Is(err, errors.NotFound) {
+			err = errors.Errorf(
+				"cannot update credential %q: controller does not manage cloud %q",
+				id.Name, id.Cloud)
+		}
+		return nil, errors.Trace(err)
+	}
+	if s.legacyUpdater == nil {
+		return modelsResult, nil
+	}
+
+	// TODO(wallyworld) - we still manage models in mongo.
+	// This can be removed after models are in dqlite.
+	// Existing credential will become valid after this call, and
+	// the model status of all models that use it will be reverted.
+	if existingInvalid != nil && *existingInvalid != cred.Invalid && !cred.Invalid {
+		tag, err := id.Tag()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if err := s.legacyUpdater(tag); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return modelsResult, nil
+}
+
+// CheckAndRevokeCredential removes the credential after first checking that any models which use the credential
+// can still access the cloud resources. If force is true, update the credential even if there are issues
+// validating the credential.
+// TODO(wallyworld) - we need a strategy to handle changes which occur after the affected models have been read
+// but before validation can complete.
+func (s *Service) CheckAndRevokeCredential(ctx context.Context, id credential.ID, force bool) error {
+	models, err := s.modelsUsingCredential(ctx, id)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(models) != 0 {
+		opMessage := "cannot be deleted as"
+		if force {
+			opMessage = "will be deleted but"
+		}
+		s.logger.Warningf("credential %v %v it is used by model%v",
```

Comment:
Hopefully my logic comes across well here but I don't see the point in logging here.

This func can either be called with force or not. In the case we don't force a sensible error is returned that the caller can use.

Why not just returned to the caller the affected models. It can then produce a logging error message where it make sense.

Reply:
We also get all the user facing message code out of services.

Reply:
I didn't want to mess with pre-existing logic in this PR. We don't know what users might rely on this. I wanted to get the current logic moved between layers first and then we can look at what might need tweaking so that such changes don't get interwoven with the lift and shift and so make things harder to understand.

Reply:
This is a log message and a 4.0 branch. We have no contracts around this. Kill it with fire I say.

Reply:
It also gets rid of the logger dependency. We don't need another thing in the tests etc etc etc.

Reply:
I changed to Debugf an removed the checks from the tests. It's still useful to have this as debug.